crates/cargo-util-schemas/src/resolve.rs

```diff
@@ -0,0 +1,250 @@
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize, de, ser};
+use url::Url;
+
+use crate::core::{GitReference, SourceKind};
+
+/// The `Cargo.lock` structure.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EncodableResolve {
+    pub version: Option<u32>,
+    pub package: Option<Vec<EncodableDependency>>,
+    /// `root` is optional to allow backward compatibility.
+    pub root: Option<EncodableDependency>,
+    pub metadata: Option<Metadata>,
+    #[serde(default, skip_serializing_if = "Patch::is_empty")]
+    pub patch: Patch,
+}
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct Patch {
+    pub unused: Vec<EncodableDependency>,
+}
+
+impl Patch {
+    fn is_empty(&self) -> bool {
+        self.unused.is_empty()
+    }
+}
+
+pub type Metadata = BTreeMap<String, String>;
+
+#[derive(Serialize, Deserialize, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub struct EncodableDependency {
+    pub name: String,
+    pub version: String,
+    pub source: Option<EncodableSourceId>,
+    pub checksum: Option<String>,
+    pub dependencies: Option<Vec<EncodablePackageId>>,
+    pub replace: Option<EncodablePackageId>,
+}
+
+#[derive(Debug, Clone)]
+pub struct EncodableSourceId {
+    /// Full string of the source
+    pub source_str: String,
+    /// Used for sources ordering
+    pub kind: SourceKind,
+    /// Used for sources ordering
+    pub url: Url,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum EncodableSourceIdParseError {
```

Comment:
imo the move to the crate should be a copy paste. If we're making changes, that should happen before or after